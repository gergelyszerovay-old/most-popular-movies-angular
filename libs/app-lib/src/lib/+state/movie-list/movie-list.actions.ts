import {createAction, props} from '@ngrx/store';
import {MovieListEntity} from "../../services/tmdb.resource.models";

export const loadMovieList = createAction(
  '[MovieList] Load MovieList',
  props<{ page: number }>()
);

export interface LoadMovieListSuccessProps {
  movieList: MovieListEntity[],
  totalMovies: number,
  totalPages: number
}

export const loadMovieListSuccess = createAction(
  '[MovieList] Load MovieList Success',
  props<LoadMovieListSuccessProps>()
);

export const loadMovieListFailure = createAction(
  '[MovieList] Load MovieList Failure',
  props<{ errorMessage: string }>()
);

export const selectMovie = createAction(
  '[MovieList] Movie Select',
  props<{ id: number }>()
);
