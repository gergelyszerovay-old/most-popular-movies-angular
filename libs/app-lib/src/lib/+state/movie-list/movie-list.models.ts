import {MovieGenresEntity, MovieListEntity} from "../../services/tmdb.resource.models";

export interface MovieListEntityWithGenreNames extends MovieListEntity {
  _genres?: Array<MovieGenresEntity>;
}

