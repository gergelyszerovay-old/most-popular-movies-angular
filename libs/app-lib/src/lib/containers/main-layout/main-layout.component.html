<div *ngIf="isLoading$() | async" class="c-main-layout x-loading">
  <div class="spinner-border" role="status">
    <span class="sr-only">Loading...</span>
  </div>
</div>
<div *ngIf="movieGenresErrorMessage$ | async" class="c-main-layout x-genre-load-error">
  <ngb-toast class="bg-danger text-light">
    Error: Cannot load the genre data from the server!<br>
    <a (click)="onReloadGenres();  $event?.preventDefault();" href="#">Click here to retry.</a>
  </ngb-toast>
</div>
<div class="container-fluid c-main-layout x-container">
  <div class="c-main-layout--header">
    <h4>Most popular movies</h4>
  </div>
  <div class="c-main-layout--content">
    <div class="row h-100 c-main-layout--content-row">
      <app-movie-list
        (onMovieClick)="onMovieSelected($event)"
        (onPageChange)="onMovieListPageSelected($event)"
        (onReloadList)="onReloadMovies()"
        [errorMessage]="movieListErrorMessage$ | async"
        [isLoading]="movieListLoading$ | async"
        [movieList]="movieList$ | async"
        [pageSize]="pageSize"
        [page]="paginationSelectedPage$ | async"
        [paginationCollectionSize]="paginationCollectionSize$ | async"
        [selectedMovie]="selectedMovie$ | async"
        class="col-md-4 h-100"
      ></app-movie-list>
      <app-movie-detailed-information
        (onCloseClick)="onDetailsClose()" [baseUrlPoster]="baseUrlPoster"
        [class.x-component-visible]="isDetailsVisibleOnMobile"
        [movie]="selectedMovie$ | async"
        class="col-md-8 h-100"
      ></app-movie-detailed-information>
    </div>
  </div>
</div>

