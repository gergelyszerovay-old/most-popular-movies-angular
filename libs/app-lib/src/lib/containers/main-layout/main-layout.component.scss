$componentClass: '.c-main-layout';

#{$componentClass} {
  &.x-loading {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 1000;
  }

  &.x-genre-load-error {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 1000;

    a {
      color: #fff;
      font-weight: bold;
      text-decoration: underline;
    }
  }

  &.x-container {
    display: flex;
    flex-direction: column;
    height: 100vh;
    overflow: hidden;
    padding: 0;

    #{$componentClass} {
      &--header {
        border-bottom: 1px solid black;
        padding: 15px;
      }

      &--content {
        flex-grow: 1;
        position: relative;

        &-row {
          position: absolute;
          width: 100%;
          margin: 0;
        }

        app-movie-list {
          padding: 0;
          overflow: auto;
        }

        app-movie-detailed-information {
          //pointer-events: none;
          padding: 0;
          overflow-x: hidden;
        }
      }

      // &--content
    }
  }
}

@media (min-width: 768px) {
  #{$componentClass} {
    &.x-container {
      #{$componentClass} {
        &--content {
          app-movie-list {
            border-right: 1px solid black;
          }
        }
      }
    }
  }
}

@media (max-width: 767px) {
  #{$componentClass} {
    &.x-container {
      #{$componentClass} {
        &--content {
          app-movie-detailed-information {
            position: absolute;
            z-index: 0;

            &.x-component-visible {
              z-index: 2;
            }
          }

          app-movie-list {
            z-index: 1;
            background-color: #fff;
          }
        }
      }
    }
  }
}





