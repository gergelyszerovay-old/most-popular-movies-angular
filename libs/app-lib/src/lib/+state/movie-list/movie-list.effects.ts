import {Injectable} from '@angular/core';
import {Actions, createEffect, ofType} from '@ngrx/effects';
import {map} from "rxjs/operators";
import {fetch} from '@nrwl/angular';

import * as MovieListActions from './movie-list.actions';
import {TmdbResourceService} from "../../services/tmdb.resource.service";
import {HttpErrorResponse} from "@angular/common/http";
import {MovieListRequestResult} from "../../services/tmdb.resource.models";

@Injectable()
export class MovieListEffects {
  loadMovieList$ = createEffect(() =>
    this.actions$.pipe(
      ofType(MovieListActions.loadMovieList),
      fetch({
        run: (action) => {
          return this.movieListService.getMovieList(action.page).pipe(
            map((response: MovieListRequestResult) => MovieListActions.loadMovieListSuccess({
              movieList: response.results,
              totalMovies: response.total_results,
              totalPages: response.total_pages
            }))
          );
        },

        onError: (action, error: HttpErrorResponse) => {
          return MovieListActions.loadMovieListFailure({errorMessage: error.message});
        },
      })
    )
  );

  constructor(private actions$: Actions, private movieListService: TmdbResourceService) {
  }
}
