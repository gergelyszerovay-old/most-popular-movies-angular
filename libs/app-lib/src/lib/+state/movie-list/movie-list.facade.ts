import {Injectable} from '@angular/core';

import {select, Store} from '@ngrx/store';

import * as MovieListSelectors from './movie-list.selectors';
import * as MovieListActions from './movie-list.actions';
import {loadMovieList} from './movie-list.actions';
import {AppLibRootStatePartial} from "../app-lib.reducer";

@Injectable()
export class MovieListFacade {
  allMovieList$ = this.store.pipe(select(MovieListSelectors.getAllMovieList));
  totalMovies$ = this.store.pipe(select(MovieListSelectors.getTotalMovies));
  totalPages$ = this.store.pipe(select(MovieListSelectors.getTotalPages));
  page$ = this.store.pipe(select(MovieListSelectors.getPage));
  selectedMovie$ = this.store.pipe(select(MovieListSelectors.getSelectedMovie));
  errorMessage$ = this.store.pipe(select(MovieListSelectors.getErrorMessage));
  loading$ = this.store.pipe(select(MovieListSelectors.getLoading));

  // allMovieListWithGenreNames$ = this.store.pipe(select(MovieListSelectors.getAllMovieListWithGenreNames));

  selectedMovieWithGenreNames$ = this.store.pipe(select(MovieListSelectors.getSelectedMovieWithGenreNames));

  constructor(private store: Store<AppLibRootStatePartial>) {
  }

  loadMovieList(page: number) {
    this.store.dispatch(loadMovieList({page}));
  }

  selectMovie(id: number) {
    this.store.dispatch(MovieListActions.selectMovie({id}));
  }
}
