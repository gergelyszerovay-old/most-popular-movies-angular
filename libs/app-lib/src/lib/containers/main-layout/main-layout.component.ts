import {ChangeDetectionStrategy, Component, OnInit} from '@angular/core';
import {combineLatest, Observable} from "rxjs";
import {MovieListFacade} from "../../+state/movie-list/movie-list.facade";
import {environment} from '@env/environment';
import {MovieListEntity} from "../../services/tmdb.resource.models";
import {MovieGenresFacade} from "../../+state/movie-genres/movie-genres.facade";
import {map} from "rxjs/operators";

@Component({
  selector: 'app-main-layout',
  templateUrl: './main-layout.component.html',
  styleUrls: ['./main-layout.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class MainLayoutComponent implements OnInit {

  movieList$: Observable<MovieListEntity[]>;
  paginationCollectionSize$: Observable<number>;
  paginationSelectedPage$: Observable<number>;
  selectedMovie$: Observable<MovieListEntity>;
  movieListErrorMessage$: Observable<string>;
  movieListLoading$: Observable<boolean>;
  movieGenresLoading$: Observable<boolean>;
  movieGenresErrorMessage$: Observable<string>;

  isDetailsVisibleOnMobile: Boolean = false;

  baseUrlPoster: string;
  pageSize: number;

  constructor(private movieListFacade: MovieListFacade, private movieGenresFacade: MovieGenresFacade) {
    this.baseUrlPoster = environment.tmdbBaseUrlPoster;
    this.pageSize = environment.tmdbDiscoverMoviePageSize;

    this.paginationCollectionSize$ = movieListFacade.totalMovies$;
    this.paginationSelectedPage$ = movieListFacade.page$;
    this.selectedMovie$ = movieListFacade.selectedMovieWithGenreNames$;
    this.movieListErrorMessage$ = movieListFacade.errorMessage$;
    this.movieList$ = movieListFacade.allMovieList$;
    this.movieListLoading$ = movieListFacade.loading$;
    this.movieGenresLoading$ = movieGenresFacade.loading$;
    this.movieGenresErrorMessage$ = movieGenresFacade.errorMessage$;
  }

  isLoading$() {
    return combineLatest([this.movieListLoading$, this.movieGenresLoading$]).pipe(
      map(([a, b]) => {
        return a || b;
      })
    );
  }

  onDetailsClose() {
    this.isDetailsVisibleOnMobile = false;
  }

  onMovieSelected(id: number) {
    this.movieListFacade.selectMovie(id)
    this.isDetailsVisibleOnMobile = true
  }

  onMovieListPageSelected(page: number) {
    this.movieListFacade.loadMovieList(page);
  }

  onReloadGenres() {
    this.movieGenresFacade.loadMovieGenres();
  }

  onReloadMovies() {
    this.movieListFacade.loadMovieList(1);
  }

  ngOnInit(): void {
  }
}
