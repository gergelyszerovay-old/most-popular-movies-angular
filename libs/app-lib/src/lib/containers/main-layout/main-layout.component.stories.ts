/*
import { text, number, boolean } from '@storybook/addon-knobs';
import { MainLayoutComponent } from './main-layout.component';

export default {
  title: 'MainLayoutComponent'
}

export const primary = () => ({
  moduleMetadata: {
    imports: []
  },
  component: MainLayoutComponent,
  props: {
    selectedMovie: text('selectedMovie', ),
  }
})
*/
