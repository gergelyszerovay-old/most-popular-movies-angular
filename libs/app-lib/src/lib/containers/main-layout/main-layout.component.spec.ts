import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {MainLayoutComponent} from './main-layout.component';
import {MockComponent, ngMocks} from "ng-mocks";
import {MovieDetailedInformationComponent} from "../../components/movie-detailed-information/movie-detailed-information.component";
import {MovieListComponent} from "../../components/movie-list/movie-list.component";
import {MovieGenresFacade, MovieListFacade} from "@most-popular-movies-angular/app-lib";
import {StoreModule} from "@ngrx/store";
import {EffectsModule} from "@ngrx/effects";
import {NgModule} from "@angular/core";
import {NxModule} from "@nrwl/angular";
import {HttpClientModule} from "@angular/common/http";
import {APPLIB_FEATURE_KEY, AppLibReducer} from "../../+state/app-lib.reducer";
import {NgbModule} from "@ng-bootstrap/ng-bootstrap";

describe('MainLayoutComponent', () => {
  let component: MainLayoutComponent;
  let fixture: ComponentFixture<MainLayoutComponent>;
  let facade: MovieListFacade;

  beforeEach(async(() => {
    @NgModule({
      imports: [
        StoreModule.forFeature(APPLIB_FEATURE_KEY, AppLibReducer)
      ],
      providers: [
        MovieListFacade,
        MovieGenresFacade
      ],
    })
    class CustomFeatureModule {
    }

    @NgModule({
      imports: [
        NxModule.forRoot(),
        StoreModule.forRoot({}),
        EffectsModule.forRoot([]),
        CustomFeatureModule,
        // AppLibModule,
        HttpClientModule
      ],
    })
    class RootModule {
    }

    TestBed.configureTestingModule({
      imports: [
        RootModule,
        NgbModule
      ],
      declarations: [
        MainLayoutComponent,
        MockComponent(MovieDetailedInformationComponent),
        MockComponent(MovieListComponent),
      ],
    })
      .compileComponents();

  }));

  beforeEach(() => {
    facade = TestBed.inject(MovieListFacade);

    fixture = TestBed.createComponent(MainLayoutComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should the snapshot match', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('@Output tests of app-movie-list', () => {
    it('onMovieSelected', () => {
      const mockedMovieListComponent = ngMocks.find<MovieListComponent>(fixture.debugElement, 'app-movie-list').componentInstance;

      spyOn(facade, 'selectMovie')

      mockedMovieListComponent.onMovieClick.emit(2)

      expect(facade.selectMovie).toHaveBeenCalledTimes(1);
      expect(facade.selectMovie).toHaveBeenCalledWith(2);
    });

    it('onMovieListPageSelected', () => {
      const mockedMovieListComponent = ngMocks.find<MovieListComponent>(fixture.debugElement, 'app-movie-list').componentInstance;

      spyOn(facade, 'loadMovieList')

      mockedMovieListComponent.onPageChange.emit(4)

      expect(facade.loadMovieList).toHaveBeenCalledTimes(1);
      expect(facade.loadMovieList).toHaveBeenCalledWith(4);
    });
  });

  it('onDetailsClose should change isDetailsVisibleOnMobile to false', () => {
    component.isDetailsVisibleOnMobile = true;
    component.onDetailsClose();
    expect(component.isDetailsVisibleOnMobile).toBe(false);
  });
});
