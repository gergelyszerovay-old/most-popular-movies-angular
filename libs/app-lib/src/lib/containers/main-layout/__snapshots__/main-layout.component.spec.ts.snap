// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayoutComponent should the snapshot match 1`] = `
<app-main-layout
  baseUrlPoster={[Function String]}
  isDetailsVisibleOnMobile="false"
  movieGenresErrorMessage$={[Function Store]}
  movieGenresFacade={[Function MovieGenresFacade]}
  movieGenresLoading$={[Function Store]}
  movieList$={[Function Store]}
  movieListErrorMessage$={[Function Store]}
  movieListFacade={[Function MovieListFacade]}
  movieListLoading$={[Function Store]}
  pageSize={[Function Number]}
  paginationCollectionSize$={[Function Store]}
  paginationSelectedPage$={[Function Store]}
  selectedMovie$={[Function Store]}
>
  <div
    class="container-fluid c-main-layout x-container"
  >
    <div
      class="c-main-layout--header"
    >
      <h4>
        Most popular movies
      </h4>
    </div>
    <div
      class="c-main-layout--content"
    >
      <div
        class="row h-100 c-main-layout--content-row"
      >
        <app-movie-list
          class="col-md-4 h-100"
        />
        <app-movie-detailed-information
          class="col-md-8 h-100"
        />
      </div>
    </div>
  </div>
</app-main-layout>
`;
